lang/py/src/avro/io.py

```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
   elif schema_type == 'int':
     if hasattr(expected_schema, 'logical_type'):
       if expected_schema.logical_type == constants.DATE:
-        return isinstance(datum, datetime.date)
+        return ValidationResult(
+          None if isinstance(datum, datetime.date) else \
+          BASE_ERR.format('logicalType(Date), baseType(datetime.date)', type(datum))
+        )
       elif expected_schema.logical_type == constants.TIME_MILLIS:
-        return isinstance(datum, datetime.time)
-    return (isinstance(datum, (int, long))
-            and INT_MIN_VALUE <= datum <= INT_MAX_VALUE)
+        return ValidationResult(
+          None if isinstance(datum, datetime.time) else \
+          BASE_ERR.format('logicalType(Time-Millis), baseType(datetime.time)', type(datum))
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and INT_MIN_VALUE <= datum <= INT_MAX_VALUE) else \
+      BASE_ERR.format('int', type(datum))
+    )
   elif schema_type == 'long':
     if hasattr(expected_schema, 'logical_type'):
       if expected_schema.logical_type == constants.TIME_MICROS:
-        return isinstance(datum, datetime.time)
+        return ValidationResult(
+          None if isinstance(datum, datetime.time) else \
+          BASE_ERR.format('logicalType(Time-Micros), baseType(datetime.time)', type(datum))
+        )
       elif expected_schema.logical_type in [constants.TIMESTAMP_MILLIS, constants.TIMESTAMP_MICROS]:
-        return isinstance(datum, datetime.datetime) and _is_timezone_aware_datetime(datum)
-    return (isinstance(datum, (int, long))
-            and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE)
+        return ValidationResult(
+          None if isinstance(datum, datetime.datetime) and _is_timezone_aware_datetime(datum) else \
+          BASE_ERR.format(
+            'logicalType(Timestamp-Millis, Timestamp-Micros), baseType(datetime.datetime)',
+            type(datum)
+          )
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE) else \
+      BASE_ERR.format('long', type(datum))
+    )
   elif schema_type in ['float', 'double']:
-    return isinstance(datum, (int, long, float))
+    return ValidationResult(
+      None if isinstance(datum, (int, long, float)) else \
+      BASE_ERR.format('(float, double)', type(datum))
+    )
   # Check for int, float, long and decimal
   elif schema_type == 'fixed':
     if (hasattr(expected_schema, 'logical_type') and
         expected_schema.logical_type == constants.DECIMAL):
-      return isinstance(datum, Decimal)
-    return isinstance(datum, str) and len(datum) == expected_schema.size
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
+        BASE_ERR.format('logicalType(Decimal), baseType(fixed)', type(datum))
+      )
+    if not isinstance(datum, str):
+      return ValidationResult(BASE_ERR.format('str', type(datum)))
```

Comment:
Is this supposed to use "fixed" type instead of "str" type since other places use the schema type in the error message? Or maybe mentioning both?